kernel/src/transaction.rs

```diff
@@ -1,19 +1,43 @@
+use std::collections::HashMap;
 use std::iter;
-use std::sync::Arc;
+use std::sync::{Arc, LazyLock};
 use std::time::{SystemTime, UNIX_EPOCH};
 
-use crate::actions::get_log_commit_info_schema;
+use crate::actions::schemas::{GetNullableContainerStructField, GetStructField};
 use crate::actions::COMMIT_INFO_NAME;
+use crate::actions::{get_log_add_schema, get_log_commit_info_schema};
 use crate::error::Error;
-use crate::expressions::{column_expr, Scalar, StructData};
+use crate::expressions::{column_expr, ColumnName, Scalar, StructData};
 use crate::path::ParsedLogPath;
-use crate::schema::{StructField, StructType};
+use crate::schema::{SchemaRef, StructField, StructType};
 use crate::snapshot::Snapshot;
 use crate::{DataType, DeltaResult, Engine, EngineData, Expression, Version};
 
+use itertools::chain;
+use url::Url;
+
 const KERNEL_VERSION: &str = env!("CARGO_PKG_VERSION");
 const UNKNOWN_OPERATION: &str = "UNKNOWN";
 
+// FIXME: should be a projection of LOG_SCHEMA?
```

Comment:
I don't think it's worth it if it's a subset of a particular action's fields.

Reply:
cool - removed!